tools/codegen/selective_build/selector.py

```diff
@@ -0,0 +1,112 @@
+from typing import Dict, Optional, Any, Set
+import yaml
+from abc import ABC, abstractmethod
+
+from tools.codegen.model import OperatorName
+from tools.codegen.selective_build.operator import SelectiveBuildOperator
+
+class SelectiveBuildBase(ABC):
+    @abstractmethod
+    def is_operator_selected(self, name: OperatorName) -> bool:
+        pass
+
+    @abstractmethod
+    def is_operator_selected_for_training(self, name: OperatorName) -> bool:
+        pass
+
+    @abstractmethod
+    def is_root_operator(self, name: OperatorName) -> bool:
+        pass
+
+
+class SelectiveBuildV1(SelectiveBuildBase):
```

Comment:
I'm really happy we started formally modeling the selective build config! Thanks!

Here are some high level thoughts:

Currently there are two forms of selective build configs:
1) OSS custom build: a simple yaml file produced by `torch.jit.export_opnames()`: https://pytorch.org/mobile/android/#custom-build
2) FB internal selective build: derived from internal BUCK rules, by merging and deduping all "pt_operator_library()" used in an App.

How are these configs used today?
They are parsed and processed by some python scripts to produce allowlists, which are sent to different pytorch codegen scripts as CLI parameters. These codegen scripts need to parse the paraemters by itself and deal with subtle semantic differences (overload v.s. no overload and etc) at different places. So changing to pass config as parsed data objects is so much better!

How is OSS build flow affected?
Please note that we need think about both OSS and FB build flows when making this type of changes. E.g. if we want to introduce the version, then we need change the `torch.jit.export_opnames()` to produce the compatible configs as well - or adding steps to do necessary conversion & updating the OSS workflow.

Shall we introduce "v1"/"v2" to OSS config?
Today the large portion of FB build flow is not in OSS repo yet, where we have more freedom to introduce new concepts like `v1`, 'v2', 'all' and etc, because we can easily migrate all call-sites if we want to change the schema. But if we introduce these to OSS then we will need be more transparent about their semantics and clarify how long we will support each and etc.

So shall we try to avoid introducing 2 versions immediately and make some effort to make the config schema more future proof?
E.g. if the main driver to introduce 2 versions is to differentiate whether omitted overload means empty "" or wildcard "*", can we explicitly model it as a property of the entry? We don't have persisted build config internally, so we can easily make the pt_operator_library() to produce the new schema programmatically. I think many incrementally added new selective build features (scalar type selection) can be optional fields, too.

Having a version in config is probably good anyway, but shall we make the version a property instead of top-level key?
E.g. When do we need have both versions in one config? Can they be separate files for internal build (as will be produced/updated separately anyway)? How will a single build support a mix of v1 & v2 configs?